drivers/codec/wm8960.py

```diff
@@ -0,0 +1,940 @@
+        primary=False,
+        adc_sync=sync_dac,
+        protocol=bus_I2S,
+        i2c_address=_WM8960_I2C_ADDR,
+    ):
+        self.i2c = i2c
+        self.i2c_address = i2c_address
+        self.value_buffer = bytearray(2)
+        self.sample_rate = sample_rate
+
+        # check parameter consistency and set the sysclk value
+        if sysclk_source == sysclk_PLL:
+            if sample_rate in (11025, 22050, 44100):
+                sysclk = 11289600
+            else:
+                sysclk = 12288000
+            if sysclk < sample_rate * 256:
+                sysclk = sample_rate * 256
+            if mclk_freq is None:
+                mclk_freq = sysclk
+        else:  # sysclk_source == sysclk_mclk
+            if mclk_freq is None:
+                mclk_freq = sample_rate * 256
+            sysclk = mclk_freq
+
+        # Reset the codec
+        self.write_reg(_WM8960_RESET, 0x00)
+        #
+        # VMID=50K, Enable VREF, AINL, AINR, ADCL and ADCR
+        # I2S_IN (bit 0), I2S_OUT (bit 1), DAP (bit 4), DAC (bit 5), ADC (bit 6) are powered on
+        self.write_reg(_WM8960_POWER1, 0xFE)
+        #
+        # Enable DACL, DACR, LOUT1, ROUT1, PLL down, SPKL, SPKR
+        self.write_reg(_WM8960_POWER2, 0x1F8)
+        #
+        # Enable left and right channel input PGA, left and right output mixer
+        self.write_reg(_WM8960_POWER3, 0x3C)
+
+        if adc_sync == sync_adc:
+            # ADC and DAC use different Frame Clock Pins
+            self.write_reg(_WM8960_IFACE2, 0x00)  # ADCLRC 0x00:Input 0x40:output.
+        else:
+            # ADC and DAC use the same Frame Clock Pin
+            self.write_reg(_WM8960_IFACE2, 0x40)  # ADCLRC 0x00:Input 0x40:output.
+        # set data route
+        self.set_data_route(route)
+        # set data protocol
+        self.set_protocol(protocol)
+
+        if sysclk_source == sysclk_PLL:
+            self.set_internal_pll_config(mclk_freq, sysclk)
+        # set master or slave
+        if primary:
+            self.set_master_clock(sysclk, sample_rate, bits)
+
+        self.set_primary(primary)
+        # set speaker clock
+        self.set_speaker_clock(sysclk)
+
+        # swap channels
+        if swap & swap_input:
+            self.modify_reg(
+                _WM8960_IFACE1,
+                _WM8960_IFACE1_ALRSWAP_MASK,
+                _WM8960_IFACE1 << _WM8960_IFACE1_ALRSWAP_SHIFT
+            )
+        if swap & swap_output:
+            self.modify_reg(
+                _WM8960_IFACE1,
+                _WM8960_IFACE1_DLRSWAP_MASK,
+                _WM8960_IFACE1 << _MM8960_IFACE1_DLRSWAP_SHIFT
+            )
+
+        # select left input
+        self.set_left_input(left_input)
+        # select right input
+        self.set_right_input(right_input)
+
+        self.write_reg(_WM8960_ADDCTL1, 0x0C0)
+        self.write_reg(_WM8960_ADDCTL4, 0x60)  # Set GPIO1 to 0.
+
+        self.write_reg(_WM8960_BYPASS1, 0x0)
+        self.write_reg(_WM8960_BYPASS2, 0x0)
```

Comment:
Since Python is a high-level language, I can't help but think there might be ways to make this code simpler and not so long/verbose.

Eg, since a lot of these pairs of regs are one after the other, there could be a `write_reg_pair` method that writes 2 at once.

And you could make a small class to manage reg read/writes with more efficient notation like subscript.   Eg, schematically:
```python
class Regs:
    def __getitem__(self, reg):
        return self.cache[reg]
    def __setitem__(self, reg, value):
        self.cache[reg] = value
        i2c.writeto_mem(addr, reg, value)

...

# use it like this
r = self.regs
r[_WM8960_BYPASS1] = 0x0
```
```

Reply:
Since reg is a list, `self.read_reg(a)` could already now be written as `self.reg[a]`, but I kept it as a function for symmetry.